# DNA Sequences Process

In this mini project, **I wrote a Python program that takes as input a file containing DNA sequences in multi-FASTA format, and made queries**. 

In [1]:
import pandas as pd
import numpy as np

## 1. Load and Prepare Data

In [2]:
# file to open
filename = 'dna.example.fasta'


try:
    fhandle = open(filename)
except IOError:
    print('The file %s does not exist !!' % filename)

In [3]:
# Put sequences in the dictionary with key=
seqs = {}

for line in fhandle:
    # let's discard the newline at the end (if any)
    line = line.rstrip()
    #print(line)
    
    #distinguish header from sequence
    if line[0] == '>': 
        words = line.split()
        name = words[0][1:]
        seqs[name]=''
    else: # sequence, not header
        seqs[name] = seqs[name] + line    

Note that a record in a FASTA file is defined as a single-line header, followed by lines of sequence data. The header line is distinguished from the sequence data by a greater-than (">") symbol in the first column. The word following the ">" symbol is the identifier of the sequence, and the rest of the line is an optional description of the entry. There should be no space between the ">" and the first letter of the identifier.

In [4]:
# Check how many records are in the file
print("There are {0} records in {1}.".format(len(seqs), filename))

There are 25 records in dna.example.fasta.


In [5]:
# Check identifiers
seqs.keys()

dict_keys(['gi|142022655|gb|EQ086233.1|43', 'gi|142022655|gb|EQ086233.1|160', 'gi|142022655|gb|EQ086233.1|41', 'gi|142022655|gb|EQ086233.1|221', 'gi|142022655|gb|EQ086233.1|294', 'gi|142022655|gb|EQ086233.1|323', 'gi|142022655|gb|EQ086233.1|564', 'gi|142022655|gb|EQ086233.1|521', 'gi|142022655|gb|EQ086233.1|455', 'gi|142022655|gb|EQ086233.1|229', 'gi|142022655|gb|EQ086233.1|422', 'gi|142022655|gb|EQ086233.1|384', 'gi|142022655|gb|EQ086233.1|280', 'gi|142022655|gb|EQ086233.1|158', 'gi|142022655|gb|EQ086233.1|59', 'gi|142022655|gb|EQ086233.1|319', 'gi|142022655|gb|EQ086233.1|438', 'gi|142022655|gb|EQ086233.1|210', 'gi|142022655|gb|EQ086233.1|237', 'gi|142022655|gb|EQ086233.1|507', 'gi|142022655|gb|EQ086233.1|350', 'gi|142022655|gb|EQ086233.1|245', 'gi|142022655|gb|EQ086233.1|279', 'gi|142022655|gb|EQ086233.1|378', 'gi|142022655|gb|EQ086233.1|101'])

Convert the dictionary `seqs`, which record DNA sequence for each individual identifier into dataframe, called `myseq`.

In [6]:
myseq = pd.Series(seqs, name="DNA_Sequence").rename_axis('ID').reset_index()

In [7]:
myseq.head()

ID  \
0   gi|142022655|gb|EQ086233.1|43   
1  gi|142022655|gb|EQ086233.1|160   
2   gi|142022655|gb|EQ086233.1|41   
3  gi|142022655|gb|EQ086233.1|221   
4  gi|142022655|gb|EQ086233.1|294   

                                        DNA_Sequence  
0  TCGGGCGAAGGCGGCAGCAAGTCGTCCACGCGCAGCGCGGCACCGC...  
1  ATTGGGGAGGAGGCGAGTTGAGCGGCGGCAGTTCGCCTGCGTGCGC...  
2  GACCTTGCATCGGCTGATCGCCGAGCGTGCCGACGTATTCATTCAC...  
3  GCCCGGCGATTTGACGTCTGCAGCCTCACGTCCAAACTCAATTTGA...  
4  GATCAGCCCCGCATACGCGTACGCGCGTGCGACGCCGAAGAGCGTC...

## 2. Sequence Analysis

### A. What are the lengths of the sequences?

In [8]:
myseq['length'] = myseq['DNA_Sequence'].apply(lambda x: len(x))

In [9]:
myseq.head()

ID  \
0   gi|142022655|gb|EQ086233.1|43   
1  gi|142022655|gb|EQ086233.1|160   
2   gi|142022655|gb|EQ086233.1|41   
3  gi|142022655|gb|EQ086233.1|221   
4  gi|142022655|gb|EQ086233.1|294   

                                        DNA_Sequence  length  
0  TCGGGCGAAGGCGGCAGCAAGTCGTCCACGCGCAGCGCGGCACCGC...     990  
1  ATTGGGGAGGAGGCGAGTTGAGCGGCGGCAGTTCGCCTGCGTGCGC...     724  
2  GACCTTGCATCGGCTGATCGCCGAGCGTGCCGACGTATTCATTCAC...    3080  
3  GCCCGGCGATTTGACGTCTGCAGCCTCACGTCCAAACTCAATTTGA...    2863  
4  GATCAGCCCCGCATACGCGTACGCGCGTGCGACGCCGAAGAGCGTC...    3832

In [10]:
# Check whether the above calculation is correct or not
assert(myseq.loc[0,'length'] == len(seqs[myseq.loc[0,'ID']]))
assert(myseq.loc[1,'length'] == len(seqs[myseq.loc[1,'ID']]))

Find the longest sequence:

In [11]:
myseq.sort_values(by='length', ascending=False).head()

ID  \
5   gi|142022655|gb|EQ086233.1|323   
14   gi|142022655|gb|EQ086233.1|59   
4   gi|142022655|gb|EQ086233.1|294   
11  gi|142022655|gb|EQ086233.1|384   
16  gi|142022655|gb|EQ086233.1|438   

                                         DNA_Sequence  length  
5   ACGCCCGGCGCACCGCGAGTACCGCGCCGCCGGGCACTCCTTGACC...    4805  
14  CGTGAATTCGATCTGCGGCAGCACGATCGCATGCATCACGCCCCGC...    4745  
4   GATCAGCCCCGCATACGCGTACGCGCGTGCGACGCCGAAGAGCGTC...    3832  
11  CTTCCGATCGCATGTCACGCCGGCGCGACATTTCGACGAACGCGTC...    3603  
16  CGGGGTCCTGATCCGGGAGTCCGGGCCGCCGGCGTATCTTCGCGTG...    3424

As one can see, the identifier, 'gi|142022655|gb|EQ086233.1|323', has the longest length of DNA sequence, 4805. 

Similarly, we can find the shortest sequence:

In [12]:
myseq.sort_values(by='length').head()

ID  \
7   gi|142022655|gb|EQ086233.1|521   
23  gi|142022655|gb|EQ086233.1|378   
8   gi|142022655|gb|EQ086233.1|455   
1   gi|142022655|gb|EQ086233.1|160   
0    gi|142022655|gb|EQ086233.1|43   

                                         DNA_Sequence  length  
7   CGTTGTTCGCCAGGTCGTCCGCATAGCCGGCCGAGCTGAACTGCGT...     512  
23  CATGGCGCGCCTGCTGCACGCGCAACGCCTGTGCGATCCAGAGCGC...     555  
8   TTCGAGTCGCTCGACGCGCCTCGACGGGCAACTAGCCGCCGTTGGC...     691  
1   ATTGGGGAGGAGGCGAGTTGAGCGGCGGCAGTTCGCCTGCGTGCGC...     724  
0   TCGGGCGAAGGCGGCAGCAAGTCGTCCACGCGCAGCGCGGCACCGC...     990

 The identifier, 'gi|142022655|gb|EQ086233.1|521', has the shortest length of DNA sequence, 512. 

### B. Find Open Reading Frame (ORF)

In molecular biology, a **reading frame** is a way of dividing the DNA sequence of nucleotides into a set of consecutive, non-overlapping triplets, called **codons**. An **open reading frame (ORF)** is the part of  reading frame that has the potential to encode a protein.  <font color='blue'>It starts with a start codon (ATG), and ends with a stop codon (TAA, TAG, or TGA) as shown in below figure, originated from [this website](https://rody.blog/2015/11/14/orf-finder-for-fasta-files-with-dna-sequences-using-python/). 
    
![alt text](https://rodybio.files.wordpress.com/2015/11/orf.jpg)    
    
<font color='black'>


Depending on where we start, there are six possible reading frames: three in the forward direction (5' to 3') and three in the reverse (3' to 5'). For instance, the three possible forward reading frames for the sequence 'AGGTGACACCGCAAGCCTTATATTAGC' are:
```
AGG TGA CAC CGC AAG CCT TAT ATT AGC  (reading frame = 1)
A GGT GAC ACC GCA AGC CTT ATA TTA GC (reading frame = 2)
AG GTG ACA CCG CAA GCC TTA TAT TAG C (reading frame = 3)
```
 For instance, <font color='red'>ATG<font color='black'>AAA<font color='red'>TAG<font color='black'> is an ORF of length 9. Proteins are formed from ORF. By analyzing the ORF we can predict the possible amino acids that might be produced during translation. [The ORF finder](https://www.ncbi.nlm.nih.gov/orffinder/) is a program available at NCBI website. It identifies all ORF or possible protein coding region from six different reading frame. In what follows, I will implement my own ORF finder written in Python.

Note: one can refer to this [website](https://vlab.amrita.edu/index.php?sub=3&brch=273&sim=1432&cnt=1) for more details about ORF.

### Self-Written ORF Finder in Python

In [13]:
def orf_finder(dna, frame=0, forward=True):
    """
    
    This function finds open reading frame (ORF) in a given DNA sequence.
    
    Parameters
    ---------
    dna: str A DNA sequence
    
    frame: int frame argument equal to 0, 1, or 2.
                   Default is 0.
    
    forward: boolean Search ORF in forward direction or not.
    
    Returns
    -------
    found_orf: list of dictionaries Return information of found ORFs
                                    including frame, start position, stop position,
                                    length of ORF, nucleotide sequence of ORF

    """
    import dnautil
    
    size = len(dna)
    
    start_codon = 'atg'
    stop_codons = ['tga', 'tag', 'taa']
    start_pos = None
    found_orf = []
    
    # forward direction (5' to 3')
    if forward:
        
        for i in range(frame, size, 3):
                
            codon = dna[i:i+3].lower()
            # Check whether codon is a start codon
            if (start_pos is None) and (codon == start_codon):
                start_pos = i
                continue
                
            # Check whether codon is a stop codon
            if (start_pos is not None) and (codon in stop_codons):
                # Record the initial position, end position, and length of ORF
                # Extract the DNA sequence by dna[start_pos:(i+3)]
                found_orf.append({"Frame": '+'+str(frame+1), "Start":start_pos+1, "Stop":i+3,\
                                 "Length": i+3 - start_pos, "ORF": dna[start_pos:i+3]})
                start_pos = None
                continue
    
    # the reverse (3' to 5') direction
    else:
        rev_complement_dna = dnautil.reverse_complement(dna)
        
        for i in range(frame, size, 3):
                
            codon = rev_complement_dna[i:i+3].lower()
            # Check whether codon is a start codon
            if (start_pos is None) and (codon == start_codon):
                start_pos = i
                continue
                
            # Check whether codon is a stop codon
            if (start_pos is not None) and (codon in stop_codons):
                # Record the initial position, end position, and length of ORF
                # Extract the DNA sequence by rev_complement_dna[start_pos:(i+3)]
                found_orf.append({"Frame": '-'+str(frame+1), "Start":size - start_pos, "Stop":size - i-2,\
                                 "Length": i+3 - start_pos, "ORF": rev_complement_dna[start_pos:i+3]})
                start_pos = None
                continue
                
        
    return found_orf

### Test `orf_finder` I wrote
In order to examine whether the code I wrote is correct or not, I compared the results with results obtained by [the ORF finder](https://www.ncbi.nlm.nih.gov/orffinder/) at NCBI website. *Note that as a matter of definition, frame i in my code is identical to frame i+1 in ORF finder at NCBI website. *

In [14]:
test_seq = seqs['gi|142022655|gb|EQ086233.1|43']   
test_seq

'TCGGGCGAAGGCGGCAGCAAGTCGTCCACGCGCAGCGCGGCACCGCGGGCCTCTGCCGTGCGCTGCTTGGCCATGGCCTCCAGCGCACCGATCGGATCAAAGCCGCTGAAGCCTTCGCGCATCAGGCGGCCATAGTTGGCGCCAGTGACCGTACCAACCGCCTTGATGCGGCGCTCGGTCATCGCTGCATTGATCGAGTAGCCACCGCCGCCGCAAATGCCCAGCACGCCAATGCGTTCTTCATCCACATAGGGGAGCGTTACGAGGTAGTCGCAGACCACGCGGAAATCCTCGACGCGCAGTGTCGGGTCTTCGGTAAAACGTGGTTCGCCGCCGCTGGCACCCTGGAAGCTGGCGTCGAAGGCGATGACGACGAAACCTTCCTTGGCCAGCGCCTCGCCATACACGTTCCCCGATGTTTGCTCCTTGCAGCTGCCGATCGGATGCGCGCTGATGATGGCGGGATATTTCTTGCCTTCGTCGAAGTTCGGCGGGAAGTGGATGTCGGCTGCGATATCCCAATACACATTCTTGATCTTGACGCTTTTCATGACAGCTCCGTTCAGGGGGAGGGGGTAAGTTCGCCAGGCCGAATCGTTGGTAGCCAAGCGGCAACGACTCGAATATAGAGAGCCGATTGGAATTCCGTAAGATCGCAATCTGGACTACAGTGGTATCTTCAAATTGACAATGGCACCTACATGGATCCCTCACTGCTTCCGTCTCTCGCGTGGTTCGCCCACGTCGCACATCATCGTAGCTTCACGAAAGCGGCTGCGGAAATGGGCGTTTCTCGAGCAAACCTGTCGCAGAACGTGAAGGCGCTCGAACGCCGGTTGAACGTCAAGCTGCTGTATCGAACGACTCGCGACATGTCGCTGACCGAGGAGGGGCAGCGGCTCTACGAGGTGTGGTATCCCGCGCTGGTCGCGGTCGAGCGGACGGTCGACGCGCTGCACGAGGAGCGCGACGAGCCGTCGGGGTTGATTC'

In [15]:
len(test_seq)

990

First of all, I consider the forward direction:

In [16]:
orf_finder(test_seq, frame=0)

[{'Frame': '+1',
  'Start': 73,
  'Stop': 135,
  'Length': 63,
  'ORF': 'ATGGCCTCCAGCGCACCGATCGGATCAAAGCCGCTGAAGCCTTCGCGCATCAGGCGGCCATAG'},
 {'Frame': '+1',
  'Start': 166,
  'Stop': 201,
  'Length': 36,
  'ORF': 'ATGCGGCGCTCGGTCATCGCTGCATTGATCGAGTAG'},
 {'Frame': '+1',
  'Start': 217,
  'Stop': 252,
  'Length': 36,
  'ORF': 'ATGCCCAGCACGCCAATGCGTTCTTCATCCACATAG'},
 {'Frame': '+1',
  'Start': 367,
  'Stop': 579,
  'Length': 213,
  'ORF': 'ATGACGACGAAACCTTCCTTGGCCAGCGCCTCGCCATACACGTTCCCCGATGTTTGCTCCTTGCAGCTGCCGATCGGATGCGCGCTGATGATGGCGGGATATTTCTTGCCTTCGTCGAAGTTCGGCGGGAAGTGGATGTCGGCTGCGATATCCCAATACACATTCTTGATCTTGACGCTTTTCATGACAGCTCCGTTCAGGGGGAGGGGGTAA'},
 {'Frame': '+1',
  'Start': 691,
  'Stop': 819,
  'Length': 129,
  'ORF': 'ATGGCACCTACATGGATCCCTCACTGCTTCCGTCTCTCGCGTGGTTCGCCCACGTCGCACATCATCGTAGCTTCACGAAAGCGGCTGCGGAAATGGGCGTTTCTCGAGCAAACCTGTCGCAGAACGTGA'}]

In [17]:
orf_finder(test_seq, frame=1)

[{'Frame': '+2',
  'Start': 416,
  'Stop': 454,
  'Length': 39,
  'ORF': 'ATGTTTGCTCCTTGCAGCTGCCGATCGGATGCGCGCTGA'}]

In [18]:
orf_finder(test_seq, frame=2)

[{'Frame': '+3',
  'Start': 444,
  'Stop': 629,
  'Length': 186,
  'ORF': 'ATGCGCGCTGATGATGGCGGGATATTTCTTGCCTTCGTCGAAGTTCGGCGGGAAGTGGATGTCGGCTGCGATATCCCAATACACATTCTTGATCTTGACGCTTTTCATGACAGCTCCGTTCAGGGGGAGGGGGTAAGTTCGCCAGGCCGAATCGTTGGTAGCCAAGCGGCAACGACTCGAATATAG'}]

In [19]:
test_seq[701:989]

'ATGGATCCCTCACTGCTTCCGTCTCTCGCGTGGTTCGCCCACGTCGCACATCATCGTAGCTTCACGAAAGCGGCTGCGGAAATGGGCGTTTCTCGAGCAAACCTGTCGCAGAACGTGAAGGCGCTCGAACGCCGGTTGAACGTCAAGCTGCTGTATCGAACGACTCGCGACATGTCGCTGACCGAGGAGGGGCAGCGGCTCTACGAGGTGTGGTATCCCGCGCTGGTCGCGGTCGAGCGGACGGTCGACGCGCTGCACGAGGAGCGCGACGAGCCGTCGGGGTTGATT'

The ORF finder at NCBI website obtained above ORF additionally but after further examination, we can see that this case is not ORF since it ends with ATT.

Next, we consider reverse direction:

In [20]:
orf_finder(test_seq, frame=0, forward=False)

[]

In [21]:
orf_finder(test_seq, frame=1, forward=False)

[{'Frame': '-2',
  'Start': 755,
  'Stop': 711,
  'Length': 45,
  'ORF': 'ATGATGTGCGACGTGGGCGAACCACGCGAGAGACGGAAGCAGTGA'}]

In [22]:
orf_finder(test_seq, frame=2, forward=False)

[{'Frame': '-3',
  'Start': 874,
  'Stop': 752,
  'Length': 123,
  'ORF': 'ATGTCGCGAGTCGTTCGATACAGCAGCTTGACGTTCAACCGGCGTTCGAGCGCCTTCACGTTCTGCGACAGGTTTGCTCGAGAAACGCCCATTTCCGCAGCCGCTTTCGTGAAGCTACGATGA'},
 {'Frame': '-3', 'Start': 703, 'Stop': 698, 'Length': 6, 'ORF': 'ATGTAG'},
 {'Frame': '-3',
  'Start': 529,
  'Stop': 260,
  'Length': 270,
  'ORF': 'ATGTGTATTGGGATATCGCAGCCGACATCCACTTCCCGCCGAACTTCGACGAAGGCAAGAAATATCCCGCCATCATCAGCGCGCATCCGATCGGCAGCTGCAAGGAGCAAACATCGGGGAACGTGTATGGCGAGGCGCTGGCCAAGGAAGGTTTCGTCGTCATCGCCTTCGACGCCAGCTTCCAGGGTGCCAGCGGCGGCGAACCACGTTTTACCGAAGACCCGACACTGCGCGTCGAGGATTTCCGCGTGGTCTGCGACTACCTCGTAA'},
 {'Frame': '-3',
  'Start': 250,
  'Stop': 179,
  'Length': 72,
  'ORF': 'ATGTGGATGAAGAACGCATTGGCGTGCTGGGCATTTGCGGCGGCGGTGGCTACTCGATCAATGCAGCGATGA'},
 {'Frame': '-3',
  'Start': 133,
  'Stop': 122,
  'Length': 12,
  'ORF': 'ATGGCCGCCTGA'}]

After test the self-written function above, we are ready to implement it to find ORFs with given DNA sequence.

### Find ORF with Given DNA Sequence

In [23]:
def orf(dna):
    """
    
    A function to find all possible ORFs with given DNA sequence

    Parameters
    ---------
    dna: str A DNA sequence
    
    
    Returns
    -------
    all_orf: list of dictionaries Return information of all found ORFs
                                    including frame, start position, stop position,
                                    length of ORF, nucleotide sequence of ORF
    """
    
    # List to record all orfs found in given DNA sequence
    all_orf = []
    
    for frame in range(3):
        # Consider forward direction
        all_orf += orf_finder(dna, frame=frame)
        
        # Consider reverse direction
        all_orf += orf_finder(dna, frame=frame, forward=False)
        
    return all_orf

In [24]:
orf(test_seq)

[{'Frame': '+1',
  'Start': 73,
  'Stop': 135,
  'Length': 63,
  'ORF': 'ATGGCCTCCAGCGCACCGATCGGATCAAAGCCGCTGAAGCCTTCGCGCATCAGGCGGCCATAG'},
 {'Frame': '+1',
  'Start': 166,
  'Stop': 201,
  'Length': 36,
  'ORF': 'ATGCGGCGCTCGGTCATCGCTGCATTGATCGAGTAG'},
 {'Frame': '+1',
  'Start': 217,
  'Stop': 252,
  'Length': 36,
  'ORF': 'ATGCCCAGCACGCCAATGCGTTCTTCATCCACATAG'},
 {'Frame': '+1',
  'Start': 367,
  'Stop': 579,
  'Length': 213,
  'ORF': 'ATGACGACGAAACCTTCCTTGGCCAGCGCCTCGCCATACACGTTCCCCGATGTTTGCTCCTTGCAGCTGCCGATCGGATGCGCGCTGATGATGGCGGGATATTTCTTGCCTTCGTCGAAGTTCGGCGGGAAGTGGATGTCGGCTGCGATATCCCAATACACATTCTTGATCTTGACGCTTTTCATGACAGCTCCGTTCAGGGGGAGGGGGTAA'},
 {'Frame': '+1',
  'Start': 691,
  'Stop': 819,
  'Length': 129,
  'ORF': 'ATGGCACCTACATGGATCCCTCACTGCTTCCGTCTCTCGCGTGGTTCGCCCACGTCGCACATCATCGTAGCTTCACGAAAGCGGCTGCGGAAATGGGCGTTTCTCGAGCAAACCTGTCGCAGAACGTGA'},
 {'Frame': '+2',
  'Start': 416,
  'Stop': 454,
  'Length': 39,
  'ORF': 'ATGTTTGCTCCTTGCAGCTGCCGATCGGATGCGCGCTGA'},
 {'Frame': '-2'

### Find All ORFs with Given DNA Sequence in Each Sample

In [25]:
myseq['orf'] = myseq.apply(lambda row: orf(row["DNA_Sequence"]), axis=1)

In [26]:
myseq.head()

ID  \
0   gi|142022655|gb|EQ086233.1|43   
1  gi|142022655|gb|EQ086233.1|160   
2   gi|142022655|gb|EQ086233.1|41   
3  gi|142022655|gb|EQ086233.1|221   
4  gi|142022655|gb|EQ086233.1|294   

                                        DNA_Sequence  length  \
0  TCGGGCGAAGGCGGCAGCAAGTCGTCCACGCGCAGCGCGGCACCGC...     990   
1  ATTGGGGAGGAGGCGAGTTGAGCGGCGGCAGTTCGCCTGCGTGCGC...     724   
2  GACCTTGCATCGGCTGATCGCCGAGCGTGCCGACGTATTCATTCAC...    3080   
3  GCCCGGCGATTTGACGTCTGCAGCCTCACGTCCAAACTCAATTTGA...    2863   
4  GATCAGCCCCGCATACGCGTACGCGCGTGCGACGCCGAAGAGCGTC...    3832   

                                                 orf  
0  [{'Frame': '+1', 'Start': 73, 'Stop': 135, 'Le...  
1  [{'Frame': '+1', 'Start': 136, 'Stop': 444, 'L...  
2  [{'Frame': '+1', 'Start': 559, 'Stop': 864, 'L...  
3  [{'Frame': '+1', 'Start': 118, 'Stop': 141, 'L...  
4  [{'Frame': '+1', 'Start': 538, 'Stop': 546, 'L...

### Find the Longest ORF in Each Sample

In [27]:
def longest(orfs, frames):
    """
    
    A function to find the longest length of ORF with given DNA sequence
    
    Parameters
    ----------
    orfs: list of dict  ORFs found by using function orf
    
    frames: list of strings  list of interested reading frames/
            'All'            Consider all reading frames
    
    Returns
    -------
    longest_length: int the longest length of ORF with given DNA sequence
    
    """
    longest_length = None
    
    # Consider all reading frames
    if frames == 'All':
    
        for orf in range(len(orfs)):
            if longest_length is None:
                longest_length = orfs[orf]['Length']
            elif longest_length < orfs[orf]['Length']:
                longest_length = orfs[orf]['Length']
                
    # Consider part of reading frames
    else:
        for orf in range(len(orfs)):
            
            # Filter out uninterested reading frames
            if not orfs[orf]['Frame'] in frames:
                continue
            
            if longest_length is None:
                longest_length = orfs[orf]['Length']
            elif longest_length < orfs[orf]['Length']:
                longest_length = orfs[orf]['Length']
            
    return longest_length

In [28]:
# Consider all reading frames in test case
myseq['longest_len_orf'] = myseq.apply(lambda row: longest(row["orf"], frames='All'), axis=1)
myseq.head()

ID  \
0   gi|142022655|gb|EQ086233.1|43   
1  gi|142022655|gb|EQ086233.1|160   
2   gi|142022655|gb|EQ086233.1|41   
3  gi|142022655|gb|EQ086233.1|221   
4  gi|142022655|gb|EQ086233.1|294   

                                        DNA_Sequence  length  \
0  TCGGGCGAAGGCGGCAGCAAGTCGTCCACGCGCAGCGCGGCACCGC...     990   
1  ATTGGGGAGGAGGCGAGTTGAGCGGCGGCAGTTCGCCTGCGTGCGC...     724   
2  GACCTTGCATCGGCTGATCGCCGAGCGTGCCGACGTATTCATTCAC...    3080   
3  GCCCGGCGATTTGACGTCTGCAGCCTCACGTCCAAACTCAATTTGA...    2863   
4  GATCAGCCCCGCATACGCGTACGCGCGTGCGACGCCGAAGAGCGTC...    3832   

                                                 orf  longest_len_orf  
0  [{'Frame': '+1', 'Start': 73, 'Stop': 135, 'Le...              270  
1  [{'Frame': '+1', 'Start': 136, 'Stop': 444, 'L...              363  
2  [{'Frame': '+1', 'Start': 559, 'Stop': 864, 'L...              918  
3  [{'Frame': '+1', 'Start': 118, 'Stop': 141, 'L...              744  
4  [{'Frame': '+1', 'Start': 538, 'Stop': 546, 'L...             2424

In [29]:
# Check Correctness
myseq[myseq['ID']=='gi|142022655|gb|EQ086233.1|43']['longest_len_orf']

0    270
Name: longest_len_orf, dtype: int64

From the results of test case we computed before, the longest length of ORF with given DNA sequence in identifier, 'gi|142022655|gb|EQ086233.1|43', is 270. It is consistent with what we found here.

In [30]:
# Consider any forward reading frames
frames = ['+1', '+2', '+3']
myseq['longest_len_orf_forward'] = myseq.apply(lambda row: longest(row["orf"], frames=frames), axis=1)
myseq.head(2)

ID  \
0   gi|142022655|gb|EQ086233.1|43   
1  gi|142022655|gb|EQ086233.1|160   

                                        DNA_Sequence  length  \
0  TCGGGCGAAGGCGGCAGCAAGTCGTCCACGCGCAGCGCGGCACCGC...     990   
1  ATTGGGGAGGAGGCGAGTTGAGCGGCGGCAGTTCGCCTGCGTGCGC...     724   

                                                 orf  longest_len_orf  \
0  [{'Frame': '+1', 'Start': 73, 'Stop': 135, 'Le...              270   
1  [{'Frame': '+1', 'Start': 136, 'Stop': 444, 'L...              363   

   longest_len_orf_forward  
0                      213  
1                      363

In [31]:
# Check Correctness
myseq[myseq['ID']=='gi|142022655|gb|EQ086233.1|43']['longest_len_orf_forward']

0    213
Name: longest_len_orf_forward, dtype: int64

It's consistent with what we found before.

### What is the length of the longest ORF appearing in any sequence and in any forward reading frame?

In [32]:
df = myseq.sort_values(by='longest_len_orf_forward', ascending=False)
print('The length of the longest ORF appearing in any sequence and in any forward reading frame is {}.'.\
      format(df.iloc[0,5]))

The length of the longest ORF appearing in any sequence and in any forward reading frame is 1686.


### What is the length of the longest ORF appearing in reading frame 2 of any of the sequences?

In [33]:
myseq['longest_len_orf_+2'] = myseq.apply(lambda row: longest(row["orf"], frames=['+2']), axis=1)

df = myseq.sort_values(by='longest_len_orf_+2', ascending=False)
print('The length of the longest ORF appearing in reading frame 2 of any of the sequences is {}.'.\
      format(df.iloc[0,6]))

The length of the longest ORF appearing in reading frame 2 of any of the sequences is 1371.0.


### What is the length of the longest ORF appearing in reading frame 3 of any of the sequences?

In [34]:
myseq['longest_len_orf_+3'] = myseq.apply(lambda row: longest(row["orf"], frames=['+3']), axis=1)

df = myseq.sort_values(by='longest_len_orf_+3', ascending=False)
print('The length of the longest ORF appearing in reading frame 3 of any of the sequences is {0}.'.\
      format(df.iloc[0,7]))

The length of the longest ORF appearing in reading frame 3 of any of the sequences is 1608.0.


In [35]:
# Find the start position of the longest ORF appearing in reading frame 3 of any of the sequences
target_dict = df.iloc[0, 3]

In [36]:
for orf in target_dict:
    if (orf['Frame']== '+3') and (orf['Length'] == df.iloc[0,7]):
        print('The starting position of the longest ORF in reading frame 3 in any of the sequences is %d.' % orf['Start'])

The starting position of the longest ORF in reading frame 3 in any of the sequences is 141.
